aclk/https_client.c

```diff
@@ -129,118 +147,408 @@ static int parse_http_response(rbuf_t buf, http_parse_ctx *parse_ctx)
+    } else {
+        buffer_strcat(hdr, ctx->request->url);
+    }
 
-    if (payload != NULL)
-        strncat(b, payload, b_size - len);
+    buffer_strcat(hdr, " HTTP/1.1\x0D\x0A");
 
-    len = strlen(b);
+    //TODO Headers!
+    if (ctx->request->request_type != HTTP_REQ_CONNECT) {
+        buffer_sprintf(hdr, "Host: %s\x0D\x0A", ctx->request->host);
+    }
+    buffer_strcat(hdr, "User-Agent: Netdata/rocks newhttpclient\x0D\x0A");
 
-    debug(D_ACLK, "Sending HTTPS req (%zu bytes): '%s'", len, b);
-    int sock = connect_to_this_ip46(IPPROTO_TCP, SOCK_STREAM, host, 0, sport, &timeout);
+    if (ctx->request->request_type == HTTP_REQ_POST && ctx->request->payload && ctx->request->payload_size) {
+        buffer_sprintf(hdr, "Content-Length: %zu\x0D\x0A", ctx->request->payload_size);
+    }
 
-    if (unlikely(sock == -1)) {
-        error("Handshake failed");
-        goto exit_buf;
+    buffer_strcat(hdr, "\x0D\x0A");
+
+    // Send the request
+    if (https_client_write_all(ctx, hdr->buffer, hdr->len)) {
+        error("Couldn't write HTTP request header into SSL connection");
+        rc = 2;
+        goto err_exit;
     }
 
-    SSL_CTX *ctx = security_initialize_openssl_client();
-    if (ctx==NULL) {
-        error("Cannot allocate SSL context");
-        goto exit_sock;
+    if (ctx->request->request_type == HTTP_REQ_POST && ctx->request->payload && ctx->request->payload_size) {
+        if (https_client_write_all(ctx, ctx->request->payload, ctx->request->payload_size)) {
+            error("Couldn't write payload into SSL connection");
+            rc = 3;
+            goto err_exit;
+        }
     }
-    // Certificate chain: not updating the stores - do we need private CA roots?
-    // Calls to SSL_CTX_load_verify_locations would go here.
-    SSL *ssl = SSL_new(ctx);
-    if (ssl==NULL) {
-        error("Cannot allocate SSL");
-        goto exit_CTX;
+
+    // Read The Response
+    if (read_parse_response(ctx)) {
+        error("Error reading or parsing response from server");
+        rc = 4;
+        goto err_exit;
     }
-    SSL_set_fd(ssl, sock);
-    ret = SSL_connect(ssl);
-    if (ret != 1) {
-        error("SSL_connect() failed with err=%d", ret);
-        goto exit_SSL;
+
+err_exit:
+    buffer_free(hdr);
+    return rc;
+}
+
+int https_request(https_req_t *request, https_req_response_t *response) {
+    int rc = 0,ret;
+    char connect_port_str[PORT_STR_MAX_BYTES];
+
+    const char *connect_host = request->proxy_host ? request->proxy_host : request->host;
+    int connect_port = request->proxy_host ? request->proxy_port : request->port;
+    struct timeval timeout = { .tv_sec = request->timeout_s, .tv_usec = 0 };
+
+    https_req_ctx_t *ctx = callocz(1, sizeof(https_req_ctx_t));
+    ctx->req_start_time = now_realtime_sec();
+
+    ctx->buf_rx = rbuf_create(RX_BUFFER_SIZE);
+    if (!ctx->buf_rx) {
+        error("Couldn't allocate buffer for RX data");
+        rc = 20;
```

Comment:
Will you do defines or something for those error codes?

Reply:
I can change them all to 1 as error code... I was just thinking maybe in future we will want to decide what to do depending on individual error cases